### PR TITLE
Adding Signal subgraph

### LINKED ISSUE #120: signal subgraph
there are now 3 papers that due "signal subgraph"

- https://arxiv.org/abs/1804.09567
- https://arxiv.org/abs/1108.1427
- https://arxiv.org/abs/1801.07683v1

let's have at least some way of doing it here.

### PR DESCRIPTION
**Pull request by caseypw**
#### Reference Issues/PRs
Fixes #120 

#### What does this implement/fix? Explain your changes.
This adds signal subgraph estimation functionality to the graspy package. Estimation of the signal subgraph of a graph will be achievable through iterative and non-iterative screening measures described in the following [paper](https://arxiv.org/abs/1108.1427).

### CODE REVIEW

File: graspy/subgraph/itscreen.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    References
+    ----------
+    .. [1] S. Wang, C. Chen, A. Badea, Priebe, C.E., Vogelstein, J.T.  "Signal 
+        Subgraph Estimation Via Vertex Screening," arXiv: 1801.07683 [stat.ME], 2018
+    """
+
+    def __init__(self, opt):
+        super().__init__(opt=opt)
+
+    def fit(self, X, y, delta, subgraph_n_verts):
```
Comment (j1c) [inline graspy/subgraph/itscreen.py]: delta, subgraph_n_verts should be part of init.

File: graspy/subgraph/itscreen.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            # Iterate
+            ind = tmpcors > tmpq
+            ind = ind.reshape(1, len(ind))
+            subgraph_inds = subgraph_inds[ind[0]]
+            itr += 1
+
+        self.corrs = cors
+        return self
+
+    def _fit_transform(self, X, y, delta, subgraph_n_verts):
```
Comment (j1c) [inline graspy/subgraph/itscreen.py]: delta, subgraph_n_verts should be part of init.

File: graspy/subgraph/itscreen.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        self.subgraph_inds = subgraph_inds
+
+        n = len(subgraph_inds)
+        S_hat = np.zeros((M, n, n))
+        for i in range(M):
+            S_hat[i] = X[i][subgraph_inds][:, subgraph_inds]
+
+        return S_hat
+
+    def fit_transform(self, X, y, delta, subgraph_n_verts):
```
Comment (j1c) [inline graspy/subgraph/itscreen.py]: delta, subgraph_n_verts should be part of init.

File: graspy/subgraph/nonitscreen.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+        Returns
+        -------
+        self : returns an instance of self.
+        """
+
+        self.corrs = self._screen(X, y)
+        return self
+
+    def _fit_transform(self, X, y, corr_threshold):
```
Comment (j1c) [inline graspy/subgraph/nonitscreen.py]: corr_threshold should be part of init

File: graspy/subgraph/nonitscreen.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        self.subgraph_inds = subgraph_inds
+
+        n = len(subgraph_inds)
+        S_hat = np.zeros((M, n, n))
+        for i in range(M):
+            S_hat[i] = X[i][subgraph_inds][:, subgraph_inds]
+
+        return S_hat
+
+    def fit_transform(self, X, y, corr_threshold):
```
Comment (j1c) [inline graspy/subgraph/nonitscreen.py]: corr_threshold should be part of init

**Comment by caseypw** [comment]
For the proof notebook, does that need to be in the graspy files or could I just upload it to github and send you guys the link?

**Comment by caseypw** [comment]
NOTE: Proof notebook for PR [here](https://github.com/NeuroDataDesign/team_graphstats1/blob/master/Casey/SS/Subgraph_Proof_Notebook.ipynb)

**Comment by caseypw** [comment]
I edited the dependencies and now it just depends on mgc since Sambit added the mgc function to the mgc  package. 

I tried having scipy 1.4.0 as the dependency for mgc  but I am pretty sure that I cannot have scipy calculate just the test statistic: it always calculates the test statistic **and** the p-value, per the [docs](https://docs.scipy.org/doc/scipy/reference/tutorial/stats.html#multiscale-graph-correlation-mgc)

This makes the computation time much longer. My bad if I am mistaken about this @sampan501 
@bdpedigo @j1c 

Also it is now giving me an error in Travis when I just have mgc as a dependency for some reason:

"ImportError: cannot import name 'MGC'"

**Comment by sampan501** [comment]
That is likely because I haven't added the version with MGC on PyPi (still testing it)

As for the calculating just the statistic, you can just set reps = 1 and to prevent the warning from showing up use something like in https://stackoverflow.com/questions/14463277/how-to-disable-python-warnings

**Comment by caseypw** [comment]
Now has scipy and mgc 
@bdpedigo @j1c

**Comment by j1c** [comment]
In your tutorial notebooks,
1) clear all outputs. part of the point of these is for netlify to build them to see if they work properly.
2) there is a line where you just print `S_hat` and something else. Not really that useful.
3) explain the experiment. the conditions, what you are proposing and and why. It can even be simpler in the sense that you can do 2 population instead of 4 for sake of clarity.

**Comment by caseypw** [comment]
@j1c 
So forego doing demonstration with three classes and instead just do the demonstration with two classes, but explain in more detail what I am doing?

**Comment by j1c** [comment]
Yes. The demo notebook should be straight forward for a layman to read so simpler experiment is better

**Comment by caseypw** [comment]
@j1c @bdpedigo I edited the notebooks and cleared all the outputs. When I re-uploaded them, the tests failed due to a Timeout Error. I then commented out part of the code that I thought was causing the Timeout error, and when I did,  the results seen now popped up: it passed the tests but Travis has never stopped running. 

Why am I getting a Tiimeout error for my tutorial notebooks?

It said "TimeoutError in tutorials/subgraph/ItScreen.ipynb: 7:53:25 PM: Cell execution timed out"

**Comment by caseypw** [comment]
@bdpedigo @j1c I changed the test statistic in the tutorial notebooks to dcorr instead of mgc and now all the tests pass but Travis never finishes

**Comment by j1c** [comment]
So apparently Travis can’t build when you have merge conflicts. Fix them and Travis should build

**Comment by caseypw** [comment]
@j1c Im not sure what the error is. I looked at the resolve conflicts and I'm not sure what it is asking. All the versions of softwares required are the same.

**Comment by j1c** [comment]
Conflicts occur when you and someone else modified the same file(s). pull master to latest commit and merge master into your branch

**Comment by caseypw** [comment]
So I incorporated all of the changes made to graspy and made my code work in Gun's subgraph directory tree. I made a new PR [here](https://github.com/neurodata/graspy/pull/304) @j1c @bdpedigo

**Pull request closed without merging**